contrib/internal/httputil: generate all possible http interfaces for the response writer wrapper

dd-caleb opened the pull request:
This fixes: https://github.com/DataDog/dd-trace-go/issues/256

So here's the idea. We need to wrap the `WriteHeader` call because that's really the only place to get the http response code. But we also want to make sure we don't accidentally remove interface implementations on the response writer (`Flusher`, `Hijacker`, etc).

One solution would be to implement a super struct:

    w = struct {
    	http.ResponseWriter
    	http.Flusher
    	http.Pusher
    	http.CloseNotifier
    	http.Hijacker
    }{w, hFlusher, hPusher, hCloseNotifier, hHijacker}

Because of field embedding we can call the methods of each interface on the outer struct (`w.Flush() -> w.Flusher.Flush()`), which in turn makes it so that we now actually implement each of the interfaces.

Unfortunately this won't work because if we didn't already implement one of the http interfaces, we would falsely indicate to the calling code that we did. For example client code might have:

    if flusher, ok := w.(http.Flusher); ok {

Which would now return true even though we didn't actually implement the method originally.

So to overcome this problem I wrote a script which generates all the possible combinations and creates an anonymous struct for each case implementing only the interfaces that the original response writer did.

Does that make sense?

gbbr commented on contrib/internal/httputil/gen/gen.go:
```diff
@@ -0,0 +1,78 @@
+// This code is generated because we have to account for all the permutations
+// of the interfaces.
+func wrapResponseWriter(w http.ResponseWriter, span ddtrace.Span) http.ResponseWriter {
+	`)
```
Would you mind using a template here? This is too cryptic and hard to follow.

gbbr commented:
This is a great solution. I love it. But please use a template so it's easy to see and work with the generated output.

gbbr commented on contrib/internal/httputil/trace_gen.go:
```diff
@@ -0,0 +1,119 @@
+// Code generated by ./gen do not modify
```
Let's respect the standard comment heading for generated files outlined here: https://github.com/golang/go/issues/13560#issuecomment-276866852

gbbr commented on contrib/internal/httputil/gen/gen.go:
```diff
@@ -0,0 +1,36 @@
+	})
+}
+
+func combinations(strs []string, pick int) (all [][]string) {
```
I've spent quite a lot of time here trying to make sense of this algorithm. Even though it looks short and witty, and possibly a well known algorithm for computer science aficiandos, I'd expect a large percentage of people visiting this code to be like me and not understand what's going on without spending time.

Would it be possible to augment it with documentation describing in detail how it works? Or, if it's a common algorithm that has a Wiki page, simply to link to it?

I was also wondering if we could rename `strs` to `list`. "strs" makes me think "strings" which is unclear when reading. But maybe it's just me.

gbbr commented on contrib/internal/httputil/gen/gen.go:
```diff
@@ -0,0 +1,36 @@
+
+func combinations(strs []string, pick int) (all [][]string) {
+	switch pick {
+	case 0:
```
I usually like to put a comment in switch cases like this. For example:
```go
    case 0:
        // nothing to do
```

gbbr commented on contrib/internal/httputil/gen/gen.go:
```diff
@@ -0,0 +1,36 @@
+		combos = append(combos, combinations(interfaces, pick))
+	}
+
+	template.Must(template.ParseFiles("./gen.gohtml")).Execute(os.Stdout, map[string]interface{}{
```
Thanks for being patient again. And I'm sorry for being unclear. I think it would be nicer and easier to handle if both the template text and the Go code would be in the same file. Basically, at the end of the Go code we'd but a `const template=...` with a backtick, like [here](https://github.com/upspin/upspin/blob/master/version/make_version.go) or [here](https://github.com/gbbr/datadog-trace-agent/blob/master/info/make.go).

Just like in those two examples, we could also just keep this in the `httputil` folder using the `// +build !ignore` flag. All in one file: `make_responsewriter.go` (or similar). I found this worked well in the past, avoiding the need for a new package.

gbbr commented on contrib/internal/httputil/gen/gen.gohtml:
```diff
@@ -0,0 +1,39 @@
+	w = newResponseWriter(w, span)
+	switch {
+{{- range .Combinations }}
+{{- range . }}
```
Could we possibly indent all these `{{ - range . }}` and `{{- end }}` commands? It's slightly difficult to see where each starts and ends.

Also, what is this `- ` prefix? I've not used it before.

dd-caleb replied:
The `-` prefix removes whitespace before the block.

gbbr commented on contrib/internal/httputil/gen/gen.go:
```diff
@@ -0,0 +1,36 @@
+package main
```
Let's document this package and explain what it does. It shouldn't be a package level documentation, it should be the documentation of this utility, something like "This program generates the implementation of all possible combinations of the http.ResponseWriter's interface implementations [...]" or something similar. To make it clear that it's not package level documentation we should add a space between the documentation and package declaration.

dd-caleb commented:
@gbbr thanks for the review. I renamed the code generated and added some additional documentation. 

Sorry that `combinations` was unclear. It's an algorithm I have a hard time reasoning about, so I tried to add an explanation of what it's doing and how it works.

dd-caleb commented:
@gbbr I moved the combinations function to a separate, internal `lists` package so that I could add a test for it
